lib/nodes/iec61850_goose.cpp

```diff
@@ -0,0 +1,577 @@
+}
+
+void GooseNode::addSubscriber(GooseNode::InputEventContext &ctx) noexcept
+{
+	GooseSubscriber subscriber;
+	SubscriberConfig &sc = ctx.subscriber_config;
+
+	ctx.node = this;
+
+	subscriber = GooseSubscriber_create(sc.go_cb_ref.data(), NULL);
+
+	if (sc.dst_address)
+		GooseSubscriber_setDstMac(subscriber, sc.dst_address->data());
+
+	if (sc.app_id)
+		GooseSubscriber_setAppId(subscriber, *sc.app_id);
+
+	GooseSubscriber_setListener(subscriber, [] (GooseSubscriber goose_subscriber, void* event_context) {
+		auto context = static_cast<InputEventContext *> (event_context);
+		onEvent(goose_subscriber, *context);
+	}, &ctx);
+
+	GooseReceiver_addSubscriber(input.receiver, subscriber);
+}
+
+void GooseNode::createReceiver() noexcept
+{
+	destroyReceiver();
+
+	input.receiver = GooseReceiver_create();
+
+	GooseReceiver_setInterfaceId(input.receiver, input.interface_id.c_str());
+
+	for (auto& pair_key_context : input.contexts)
+		addSubscriber(pair_key_context.second);
+
+	input.state = GooseNode::Input::READY;
+}
+
+void GooseNode::destroyReceiver() noexcept
+{
+	int err __attribute__((unused));
+
+	if (input.state == GooseNode::Input::NONE)
+		return;
+
+	stopReceiver();
+
+	GooseReceiver_destroy(input.receiver);
+
+	err = queue_signalled_destroy(&input.queue);
+
+	input.state = GooseNode::Input::NONE;
+}
+
+void GooseNode::startReceiver() noexcept(false)
+{
+	if (input.state == GooseNode::Input::NONE)
+		createReceiver();
+	else
+		stopReceiver();
+
+	GooseReceiver_start(input.receiver);
+
+	if (!GooseReceiver_isRunning(input.receiver))
+		throw RuntimeError{"iec61850-GOOSE receiver could not be started"};
+
+	input.state = GooseNode::Input::READY;
+}
+
+void GooseNode::stopReceiver() noexcept
+{
+	if (input.state == GooseNode::Input::NONE)
+		return;
+
+	input.state = GooseNode::Input::STOPPED;
+
+	if (!GooseReceiver_isRunning(input.receiver))
+		return;
+
+	GooseReceiver_stop(input.receiver);
+}
+
+int GooseNode::_read(Sample *samples[], unsigned sample_count)
+{
+	int available_samples;
+	struct Sample *copies[sample_count];
+
+	if (input.state != GooseNode::Input::READY)
+		return -1;
+
+	available_samples = queue_signalled_pull_many(&input.queue, (void **) copies, sample_count);
+	sample_copy_many(samples, copies, available_samples);
+	sample_decref_many(copies, available_samples);
+
+	return available_samples;
+}
+
+GooseNode::GooseNode(const std::string &name) :
+	Node(name)
+{
+	input.enabled = false;
+	input.state = GooseNode::Input::NONE;
+
+	input.contexts = {};
+	input.mappings = {};
+	input.interface_id = "lo";
+	input.queue_length = 1024;
+}
+
+GooseNode::~GooseNode()
+{
+	int err __attribute__((unused));
+
+	destroyReceiver();
+
+	err = queue_signalled_destroy(&input.queue);
+
+	err = pool_destroy(&input.pool);
+}
+
+int GooseNode::parse(json_t *json, const uuid_t sn_uuid)
+{
+	int ret;
+	json_error_t err;
+
+	ret = Node::parse(json, sn_uuid);
+	if (ret)
+		return ret;
+
+	ret = _parse(json, &err);
```

Comment:
Please merge `_parse` into this function.

I dont see any other callsite of `_parse`. So I dont see a reason to factor it out to a dedicated function?

Reply:
I wanted to only have one site where the `throw ConfigError(json, err, "node-config-node-iec61850-goose")` occures. This allowed all parse related functions to adhere to the `if (ret) return ret;` pattern to propagate json parsing errors.

Reply:
The ` if (ret) return ret;` pattern is actually a relict from the older C code. I think we should migrate to use exceptions for configuration parsing.

It has also the advantage, that we can more clearly report where the error occurred (line number of JSON config).
Have a look at the other occurences of ConfigError. In many instances we also included an ID which we used to point the user directly to the documentation where the setting is documented.